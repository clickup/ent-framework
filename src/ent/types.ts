import type { Schema } from "../abstract/Schema";
import type {
  CountInput,
  ExistsInput,
  ID,
  LoadByInput,
  Order,
  Table,
  UniqueKey,
  Where,
} from "../types";
import type { Validation } from "./Validation";
import type { VC } from "./VC";

/**
 * A very shallow interface of Ent class (as a collection of static methods).
 * Used in some places where we need the very minimum from the Ent.
 */
export interface EntClass<TTable extends Table = any> {
  readonly SCHEMA: Schema<TTable>;
  readonly VALIDATION: Validation<TTable>;
  readonly name: string; // class constructor name

  new (): Ent;
  loadX(vc: VC, id: string): Promise<Ent>;
  loadNullable(vc: VC, id: string): Promise<Ent | null>;
  loadIfReadableNullable(vc: VC, id: string): Promise<Ent | null>;
  count(vc: VC, where: CountInput<TTable>): Promise<number>;
  exists(vc: VC, where: ExistsInput<TTable>): Promise<boolean>;
  select(
    vc: VC,
    where: Where<TTable>,
    limit: number,
    order?: Order<TTable>
  ): Promise<Ent[]>;
  selectChunked(
    vc: VC,
    where: Where<TTable>,
    chunkSize: number,
    limit: number,
    custom?: {}
  ): AsyncIterableIterator<Ent[]>;
  loadByX(vc: VC, keys: LoadByInput<TTable, UniqueKey<TTable>>): Promise<Ent>;
}

/**
 * A very shallow interface of one Ent.
 */
export interface Ent {
  readonly [ID]: string;
  readonly vc: VC;
  deleteOriginal(): Promise<boolean>;
}
