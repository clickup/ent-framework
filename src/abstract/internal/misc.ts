export const OP_TIMELINE_POS_REFRESH = "TIMELINE_POS_REFRESH";
export const OP_PING = "PING";
export const OP_SHARD_NOS = "SHARD_NOS";
