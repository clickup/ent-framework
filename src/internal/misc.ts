import { createHash } from "crypto";
import { inspect } from "util";
import compact from "lodash/compact";
import objectHashModule from "object-hash";

/**
 * In some cases (e.g. when actively working with callbacks), TS is still weak
 * enough, so we are not always able to use generics/unknown/never types.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type DesperateAny = any;

/**
 * Removes constructor signature from a type.
 * https://github.com/microsoft/TypeScript/issues/40110#issuecomment-747142570
 */
export type OmitNew<T extends new (...args: never[]) => unknown> = Pick<
  T,
  keyof T
>;

/**
 * Adds a type alternative to constructor signature's return value. This is
 * useful when we e.g. turn an instance of some Ent class into an Instance & Row
 * type where Row is dynamically inferred from the schema.
 */
export type AddNew<
  TClass extends new (...args: never[]) => unknown,
  TRet,
> = OmitNew<TClass> & { new (): InstanceType<TClass> & TRet };

/**
 * Flattens the interface to make it more readable in IntelliSense. Can be used
 * when someone modifies (picks, omits, etc.) a huge type.
 */
export type Flatten<T> = {} & { [P in keyof T]: T[P] };

/**
 * Cancels "readonly" specifier on object's properties.
 */
export type Writeable<T> = { -readonly [P in keyof T]: T[P] };

/**
 * Returns a union type of all tuple strict prefixes:
 * ["a", "b", "c"] -> ["a", "b"] | ["a"]
 */
export type TuplePrefixes<T extends readonly unknown[]> = T extends [unknown]
  ? []
  : T extends [infer First, ...infer Rest]
    ? [First, ...TuplePrefixes<Rest>] | [First]
    : [];

/**
 * Picks only partial (optional) keys of an object.
 */
export type PickPartial<T> = {
  [K in keyof T as undefined extends T[K] ? K : never]: T[K];
};

/**
 * Denotes an option which can be dynamically configured at runtime.
 */
export type MaybeCallable<T> = T | (() => T);

/**
 * Some Node APIs throw not an instance of Error object, but something looking
 * like an Error. So we can't do "instanceof Error" check in all cases, we can
 * only compare the shape of a variable received in a `catch (e: unknown)` block
 * and hope for best.
 */
export type MaybeError<TExtra extends {} = {}> =
  | ({ code?: string; message?: string; stack?: string } & Partial<TExtra>)
  | null
  | undefined;

/**
 * Turns a list of Promises to a list of Promise resolution results.
 */
export async function join<TList extends readonly unknown[]>(
  promises: TList,
): Promise<{ -readonly [P in keyof TList]: Awaited<TList[P]> }>;

/**
 * Turns an object where some values are Promises to an object with values as
 * Promise resolution results.
 */
export async function join<TRec extends Readonly<Record<string, unknown>>>(
  promises: TRec,
): Promise<{ -readonly [K in keyof TRec]: Awaited<TRec[K]> }>;

/**
 * A safe replacement for Promise-all built-in method.
 *
 * Works the same way as Promise-all, but additionally guarantees that ALL OTHER
 * promises have settled in case one of them rejects. This is needed to ensure
 * that we never have unexpected "dangling" promises continuing running in
 * nowhere in case one of the promises rejects early (the behavior of
 * Promise.all is to reject eagerly and let the rest of stuff running whilst the
 * caller code unfreezes).
 *
 * The behavior of join() is similar to Promise.allSettled(), but it throws the
 * 1st exception occurred; this is what's expected in most of the cases, and
 * this is how promises are implemented in e.g. Hack.
 *
 * The benefits of ensuring everything is settled:
 *
 * 1. We never have surprising entries in our logs (e.g. imagine a request
 *    aborted long time ago, and then some "dangling" promises continue running
 *    and issue queries as if nothing happened).
 * 2. Predictable control flow: if we run `await join()`, we know that no side
 *    effects from the spawned promises will appear after this await throws or
 *    returns.
 *
 * "Join" is a term from parallel programming (e.g. "join threads"), it’s pretty
 * concrete and means that after the call, multiple parallel execution flows
 * “join” into one. It's a word to describe having "one" from "many".
 *
 * What’s interesting is that, besides Promise-all leaks execution flows, it
 * still doesn’t trigger unhandledRejection for them in case one of them throws
 * later, it just swallows all other exceptions.
 *
 * I.e. Promise-all means "run all in parallel, if one throws - throw
 * immediately and let the others continue running in nowhere; if some of THAT
 * others throws, swallow their exceptions".
 *
 * And join() means "run all in parallel, if one throws - wait until everyone
 * finishes, and then throw the 1st exception; if some of others throw, swallow
 * their exceptions".
 *
 * See also https://en.wikipedia.org/wiki/Fork%E2%80%93join_model
 */
export async function join(promises: unknown[] | object): Promise<unknown> {
  const promisesArray =
    promises instanceof Array ? promises : Object.values(promises);
  let firstError: unknown = undefined;
  let errorCount = 0;
  const resultsArray = await Promise["all"](
    promisesArray.map(async (promise) =>
      Promise.resolve(promise).catch((err) => {
        if (errorCount === 0) {
          firstError = err;
        }

        errorCount++;
        return undefined;
      }),
    ),
  );
  if (errorCount > 0) {
    throw firstError;
  }

  return promises instanceof Array
    ? resultsArray
    : Object.fromEntries(
        Object.keys(promises).map((key, i) => [key, resultsArray[i]]),
      );
}

/**
 * A shortcut for `await join(arr.map(async ...))`.
 */
export async function mapJoin<TElem, TRet>(
  arr: readonly TElem[] | Promise<readonly TElem[]>,
  func: (e: TElem, idx: number) => PromiseLike<TRet> | TRet,
): Promise<TRet[]> {
  return join((await arr).map((e, idx) => func(e, idx)));
}

/**
 * Returns a random value between 1 and 1+jitter.
 */
export function jitter(jitter: number): number {
  return 1 + jitter * Math.random();
}

/**
 * Copies a stack-trace from fromErr error into toErr object. Useful for
 * lightweight exceptions wrapping.
 */
export function copyStack<
  TError extends Error,
  TFrom extends { stack?: unknown; message?: unknown } | null | undefined,
>(toErr: TError, fromErr: TFrom): TError {
  if (
    typeof fromErr?.stack !== "string" ||
    typeof fromErr?.message !== "string"
  ) {
    return toErr;
  }

  // This is magic, the 1st line in stacktrace must be exactly "ExceptionType:
  // exception message\n", otherwise jest goes mad and prints the stacktrace
  // incorrectly (once from err.message and then once from err.stack). See also:
  // https://stackoverflow.com/questions/42754270/re-throwing-exception-in-nodejs-and-not-losing-stack-trace
  const fromMessageLines = fromErr.message.split("\n").length;
  toErr.stack =
    toErr.toString() + // original toErr message
    "\n" +
    fromErr.stack
      .split("\n")
      .slice(fromMessageLines) // skip prefix=fromErr.message in fromErr.stack
      .join("\n");
  return toErr;
}

/**
 * Tries to minify a stacktrace by removing common parts of the paths. See unit
 * test with snapshot for examples.
 */
export function minifyStack(stack: string, framesToPop: number): string {
  return stack
    .replace(/^\w+:[ ]*\n/s, "") // remove "Error:" prefix
    .trim()
    .split("\n")
    .slice(framesToPop)
    .join("\n")
    .replace(/^\s+/gm, "")
    .replace(/^[^\n]+\(<anonymous>\)\n/gm, "")
    .replace(/(:\d+):\d+(?=[\n)])/gs, "$1")
    .replace(/^(at )\/.+\//gm, "$1")
    .replace(/^(at [^\n]+\()\/.+\//gm, "$1")
    .replace(/^(at )([^\n]+?) \((.+)\)/gm, "$1$3 ($2)");
}

/**
 * A simple sequence generator which never returns the same value twice within
 * the same process. It's NOT random, NOT for cryptography, NOT stored (so
 * starts from scratch on a process restart) and is NOT shared with other
 * processes.
 */
export function localUniqueInt(): number {
  return sequenceValue++;
}

let sequenceValue = 1;

/**
 * The quickest string hasher. Don't use for crypto purposes!
 * https://medium.com/@chris_72272/what-is-the-fastest-node-js-hashing-algorithm-c15c1a0e164e
 */
export function stringHash(s: string): string {
  return createHash("sha1").update(s).digest("hex");
}

/**
 * Used to calculate stable hashes of e.g. unique keys.
 */
export function objectHash(obj: object): Buffer {
  return objectHashModule(obj, {
    algorithm: "sha1",
    encoding: "buffer",
  });
}

/**
 * Similar to objectHash(), but uses JSON.stringify() under the hood, assuming
 * that it's faster than objectHash(). Also, doesn't throw when the object
 * contains bigint values (as opposed to JSON.stringify()).
 */
export function jsonHash(obj: unknown): string {
  return stringHash(
    JSON.stringify(obj, (_, value) =>
      typeof value === "bigint" ? value.toString() : value,
    ),
  );
}

/**
 * Indents each line of the text with 2 spaces.
 */
export function indent(message: string): string {
  return message.replace(/^/gm, "  ");
}

/**
 * Adds text suffixes to the sentence (typically, to an error message).
 */
export function addSentenceSuffixes(
  sentence: string,
  ...suffixes: Array<string | undefined>
): string {
  const compacted = compact(suffixes);
  if (compacted.length === 0) {
    return sentence;
  }

  const suffix = compacted
    .filter((suffix) => !sentence.endsWith(suffix))
    .join("");
  return suffix.startsWith("\n")
    ? sentence + suffix
    : sentence.trimEnd().replace(/[.!?]+$/s, "") + suffix;
}

/**
 * Returns the 1st line of the message.
 */
export function firstLine<T extends string | undefined>(message: T): T {
  return (
    typeof message === "string" ? message.replace(/\n.*/s, "") : message
  ) as T;
}

/**
 * A shorthand for inspect() in compact/no-break mode.
 */
export function inspectCompact(obj: unknown): string {
  return inspect(obj, { compact: true, breakLength: Infinity }).replace(
    /^([[])\s+|\s+([\]])$/gs,
    (_, $1, $2) => $1 || $2,
  );
}

/**
 * Prepares something which is claimed to be an ID for debug printing in e.g.
 * exception messages. We replace all non-ASCII characters to their \u
 * representations.
 */
export function sanitizeIDForDebugPrinting(idIn: unknown): string {
  const MAX_LEN = 32;
  const id = "" + idIn;
  const value =
    id
      .substring(0, MAX_LEN)
      // We want to use control characters in this regex.
      // eslint-disable-next-line no-control-regex
      .replace(/[^\x1F-\x7F]/g, (v) => "\\u" + v.charCodeAt(0)) +
    (id.length > MAX_LEN ? "..." : "");
  return value === "" ? '""' : value;
}

/**
 * Throws if the value passed is null or undefined.
 */
export function nullthrows<T>(
  x?: T | null,
  message?: (() => string | Error) | string | Error,
): T {
  if (x !== null && x !== undefined) {
    return x;
  }

  if (typeof message === "function") {
    message = message();
  }

  const error =
    message instanceof Error
      ? message
      : Error(message ?? `Got unexpected ${x} in nullthrows()`);
  Error.captureStackTrace(error, nullthrows);
  throw error;
}

/**
 * Two modes:
 * 1. If an async (or sync) function is passed, spawns it in background and
 *    doesn't await for its termination.
 * 2. If a Promise is passed, lets it continue executing, doesn't await on it.
 *
 * Useful when we want to launch a function "in the air", "hanging in nowhere",
 * and make no-misused-promises and no-floating-promises rules happy with it. An
 * example is some legacy callback-based API (e.g. chrome extension API) where
 * we want to pass an async function.
 *
 * It's like an analog of "async on intent" comment in the code.
 */
export function runInVoid(
  funcOrPromise: (() => Promise<unknown> | void) | Promise<unknown> | void,
): void {
  if (funcOrPromise instanceof Function) {
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    funcOrPromise();
  } else {
    // do nothing, our Promise is already hanging in nowhere
  }
}

/**
 * A typesafe-way to invariant the object's key presence and being
 * non-undefined. It is not always working for union types: sometimes it asserts
 * the value of the key to be "any". It also doesn't remove "undefined" from the
 * type of the value.
 */
export function hasKey<K extends symbol | string>(
  k: K,
  o: unknown,
): o is { [_ in K]: DesperateAny } {
  return (
    !!o &&
    (typeof o === "object" || typeof o === "function") &&
    k in o &&
    (o as Record<K, unknown>)[k] !== undefined
  );
}

/**
 * Same as Object.entries(), but returns strongly-typed entries.
 */
export function entries<K extends string, V>(
  obj: Partial<Record<K, V>>,
): Array<[K, V]> {
  return Object.entries(obj) as Array<[K, V]>;
}

/**
 * If the passed value is a function, calls it; otherwise, returns it intact.
 */
export function maybeCall<T>(valueOrFn: MaybeCallable<T>): T {
  return typeof valueOrFn === "function" || valueOrFn instanceof Function
    ? (valueOrFn as Function)()
    : valueOrFn;
}
