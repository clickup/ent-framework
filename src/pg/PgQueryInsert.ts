import type { QueryAnnotation } from "../abstract/QueryAnnotation";
import { QueryBase } from "../abstract/QueryBase";
import type { Schema } from "../abstract/Schema";
import type { InsertInput, Table } from "../types";
import { ID } from "../types";
import type { PgClient } from "./PgClient";
import { PgRunner } from "./PgRunner";

export class PgQueryInsert<TTable extends Table> extends QueryBase<
  TTable,
  InsertInput<TTable>,
  string | null,
  PgClient
> {
  /** @ignore */
  readonly RUNNER_CLASS = PgRunnerInsert;
}

class PgRunnerInsert<TTable extends Table> extends PgRunner<
  TTable,
  InsertInput<TTable>,
  string | null
> {
  static override readonly IS_WRITE = true;
  private singleBuilder;
  private batchBuilder;

  readonly op = "INSERT";
  readonly maxBatchSize = 100;
  readonly default = null; // In case of duplicate key error, returns null.

  constructor(schema: Schema<TTable>, client: PgClient) {
    super(schema, client);

    const fields = this.addPK(Object.keys(this.schema.table), "append");

    this.singleBuilder = this.createValuesBuilder({
      prefix: this.fmt("INSERT INTO %T (%FIELDS) VALUES", { fields }),
      indent: "  ",
      fields,
      suffix: this.fmt(` ON CONFLICT DO NOTHING RETURNING %PK AS ${ID}`),
    });

    // We use WITH clause in INSERT, because "ON CONFLICT DO NOTHING" clause
    // doesn't emit anything in "RETURNING" clause, so we could've not
    // distinguished rows which were inserted from the rows which were not.
    // Having WITH solves this (see RETURNING below).
    this.batchBuilder = this.createWithBuilder({
      fields,
      suffix: this.fmt(
        "INSERT INTO %T (%FIELDS)\n" +
          "SELECT %FIELDS FROM rows OFFSET 1\n" +
          `ON CONFLICT DO NOTHING RETURNING (SELECT _key FROM rows WHERE %PK(rows)=%PK(%T)), %PK AS ${ID}`,
        { fields },
      ),
    });
  }

  override key(input: InsertInput<TTable>): string {
    // We must NEVER dedup inserts, because:
    // 1. If the table DOESN'T have an unique key, then we must insert all
    //    input rows (no dedup allowed).
    // 2. If the table DOES have an unique key, then we must logically ensure
    //    that only one concurrent promise is resolved into an inserted row ID,
    //    and all other are resolved with null (aka "not inserted due to
    //    duplicate").
    return super.key(input);
  }

  async runSingle(
    input: InsertInput<TTable>,
    annotations: QueryAnnotation[],
  ): Promise<string | undefined> {
    const sql =
      this.singleBuilder.prefix +
      this.singleBuilder.func([["", input]]) +
      this.singleBuilder.suffix;
    const rows = await this.clientQuery<{ [ID]: string }>(sql, annotations, 1);
    if (!rows.length) {
      return undefined;
    }

    return rows[0][ID];
  }

  async runBatch(
    inputs: Map<string, InsertInput<TTable>>,
    annotations: QueryAnnotation[],
  ): Promise<Map<string, string>> {
    const sql =
      this.batchBuilder.prefix +
      this.batchBuilder.func(inputs) +
      this.batchBuilder.suffix;
    const rows = await this.clientQuery<{ _key: string; [ID]: string }>(
      sql,
      annotations,
      inputs.size,
    );
    const outputs = new Map<string, string>();
    for (const row of rows) {
      outputs.set(row._key, row[ID]);
    }

    return outputs;
  }
}
