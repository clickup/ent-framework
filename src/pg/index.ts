/**
 * @file Automatically generated by barrelsby.
 */

export * from "./PgClient";
export * from "./PgError";
export * from "./PgQueryCount";
export * from "./PgQueryDelete";
export * from "./PgQueryDeleteWhere";
export * from "./PgQueryExists";
export * from "./PgQueryIDGen";
export * from "./PgQueryInsert";
export * from "./PgQueryLoad";
export * from "./PgQueryLoadBy";
export * from "./PgQuerySelect";
export * from "./PgQuerySelectBy";
export * from "./PgQueryUpdate";
export * from "./PgQueryUpsert";
export * from "./PgRunner";
export * from "./PgSchema";
export * from "./PgShardNamer";
export * from "./helpers/buildShape";
export * from "./helpers/escapeIdent";
export * from "./helpers/escapeLiteral";
export * from "./types/BigIntArrayType";
export * from "./types/ByteaBufferType";
export * from "./types/StringArrayType";
