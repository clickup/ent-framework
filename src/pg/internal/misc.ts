import type {
  ClientErrorKind,
  ClientErrorPostAction,
} from "../../abstract/ClientError";

/**
 * Some errors affect the logic of choosing another Client when a query retry is
 * requested (controlled via ClientError). Mostly, those are the situation when
 * a PG node goes down.
 *
 * Notice that, if a master goes down, we always need to find another master
 * (and retry hard hoping is a switchover, and it will appear on another node).
 */
export const CLIENT_ERROR_PREDICATES: Array<
  (error: { code: string; message: string }) =>
    | false
    | {
        abbreviation: string;
        postAction: {
          ifMaster: ClientErrorPostAction;
          ifReplica: ClientErrorPostAction;
          reportConnectionIssue: boolean;
        };
        kind: ClientErrorKind;
        comment: string;
      }
> = [
  ({ code }) =>
    code === "42P01" && {
      abbreviation: "undefined_table",
      postAction: {
        ifMaster: "rediscover-cluster",
        ifReplica: "rediscover-cluster",
        reportConnectionIssue: false,
      },
      kind: "data-on-server-is-unchanged",
      comment:
        "For single-queries: table doesn't exist or disappeared (e.g. the Shard was relocated to another Island).",
    },
  ({ code }) =>
    code === "42704" && {
      abbreviation: "undefined_object",
      postAction: {
        ifMaster: "rediscover-cluster",
        ifReplica: "rediscover-cluster",
        reportConnectionIssue: false,
      },
      kind: "data-on-server-is-unchanged",
      comment:
        "For batched queries: table doesn't exist or disappeared (e.g. the Shard was relocated to another Island).",
    },
  ({ code }) =>
    code === "25006" && {
      abbreviation: "read_only_sql_transaction",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "rediscover-island",
        reportConnectionIssue: false,
      },
      kind: "data-on-server-is-unchanged",
      comment:
        "A write happened in a read-only Client (probably the Client's role was changed from master to replica due to a failover/switchover).",
    },
  ({ code }) =>
    code === "57P01" && {
      abbreviation: "admin_shutdown",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "data-on-server-is-unchanged",
      comment: "The database is shutting down by an administrator.",
    },
  ({ code }) =>
    code === "57P02" && {
      abbreviation: "crash_shutdown",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "data-on-server-is-unchanged",
      comment: "The database is crashed and is shutting down.",
    },
  ({ message }) =>
    message === "Connection terminated unexpectedly" && {
      abbreviation: "connection_terminated",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "unknown-server-state", // !!!
      comment:
        "Node-postgres connection terminated unexpectedly (from client.js).",
    },
  ({ code }) =>
    code === "ECONNREFUSED" && {
      abbreviation: "ECONNREFUSED",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "data-on-server-is-unchanged",
      comment: "Node TCP library connect error.",
    },
  ({ code, message }) =>
    code === "08P01" && // protocol_violation
    message.includes("server conn crashed") && {
      abbreviation: "server_conn_crashed",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "unknown-server-state",
      comment: "PG is terminated cruelly (e.g. by SIGKILL or SIGSEGV).",
    },
  ({ code, message }) =>
    code === "08P01" && // protocol_violation
    message === "query_wait_timeout" && {
      abbreviation: "query_wait_timeout",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "data-on-server-is-unchanged",
      comment:
        "PG went down a few seconds ago, but a PgBouncer connection is still open, so the query sent to it times out waiting. After some time, PgBouncer will stop emitting this error and start emitting server_login_retry on new connections instead.",
    },
  ({ code, message }) =>
    code === "08P01" && // protocol_violation
    message.includes("server_login_retry") && {
      abbreviation: "server_login_retry",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "data-on-server-is-unchanged",
      comment:
        "PG has been down for quite some, so PgBouncer emits this error on each new connection immediately. Or, PG server has just been shut down, and PgBouncer's server pool is empty (fast-fail use case).",
    },
  ({ code, message }) =>
    code === "08P01" && // protocol_violation
    message === "pgbouncer cannot connect to server" && {
      abbreviation: "cannot_connect_to_server",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "data-on-server-is-unchanged",
      comment:
        "Older PgBouncer versions emit this error instead of server_login_retry.",
    },
  ({ code }) =>
    code === "08P01" && {
      // protocol_violation
      abbreviation: "protocol_violation_catch_all",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "data-on-server-is-unchanged",
      comment:
        'Sometimes PgBouncer emit unrelated error codes when PG server crashes (e.g. "invalid server parameter").',
    },
];
