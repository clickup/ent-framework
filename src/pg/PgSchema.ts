import type { Query } from "../abstract/Query";
import { Schema } from "../abstract/Schema";
import type {
  CountInput,
  ExistsInput,
  InsertInput,
  LoadByInput,
  Row,
  SelectByInput,
  SelectInput,
  Table,
  UniqueKey,
  UpdateInput,
} from "../types";
import { PgQueryCount } from "./PgQueryCount";
import { PgQueryDelete } from "./PgQueryDelete";
import { PgQueryExists } from "./PgQueryExists";
import { PgQueryIDGen } from "./PgQueryIDGen";
import { PgQueryInsert } from "./PgQueryInsert";
import { PgQueryLoad } from "./PgQueryLoad";
import { PgQueryLoadBy } from "./PgQueryLoadBy";
import { PgQuerySelect } from "./PgQuerySelect";
import { PgQuerySelectBy } from "./PgQuerySelectBy";
import { PgQueryUpdate } from "./PgQueryUpdate";
import { PgQueryUpsert } from "./PgQueryUpsert";

export class PgSchema<
  TTable extends Table,
  TUniqueKey extends UniqueKey<TTable>,
> extends Schema<TTable, TUniqueKey> {
  idGen(): Query<string> {
    return new PgQueryIDGen(this);
  }

  insert(input: InsertInput<TTable>): Query<string | null> {
    return new PgQueryInsert(this, input);
  }

  upsert(input: InsertInput<TTable>): Query<string> {
    return new PgQueryUpsert(this, input);
  }

  update(id: string, input: UpdateInput<TTable>): Query<boolean> {
    return new PgQueryUpdate(this, id, input);
  }

  delete(id: string): Query<boolean> {
    return new PgQueryDelete(this, id);
  }

  load(id: string): Query<Row<TTable> | null> {
    return new PgQueryLoad(this, id);
  }

  loadBy(input: LoadByInput<TTable, TUniqueKey>): Query<Row<TTable> | null> {
    return new PgQueryLoadBy(this, input);
  }

  selectBy(
    input: SelectByInput<TTable, TUniqueKey>,
  ): Query<Array<Row<TTable>>> {
    return new PgQuerySelectBy(this, input);
  }

  select(input: SelectInput<TTable>): Query<Array<Row<TTable>>> {
    return new PgQuerySelect(this, input);
  }

  count(input: CountInput<TTable>): Query<number> {
    return new PgQueryCount(this, input);
  }

  exists(input: ExistsInput<TTable>): Query<boolean> {
    return new PgQueryExists(this, input);
  }
}
