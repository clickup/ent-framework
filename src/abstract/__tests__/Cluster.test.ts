import { MASTER } from "..";
import { join } from "../../internal/misc";
import {
  TCPProxyServer,
  TEST_CONFIG,
  testCluster,
} from "../../pg/__tests__/test-utils";
import { PgQueryInsert } from "../../pg/PgQueryInsert";
import { PgSchema } from "../../pg/PgSchema";
import { ID } from "../../types";

jest.useFakeTimers({ advanceTimers: true });

let connStuckServer: TCPProxyServer;

beforeEach(async () => {
  connStuckServer = new TCPProxyServer({
    host: TEST_CONFIG.host!,
    port: TEST_CONFIG.port!,
    delayOnConnect: 1000000,
  });
});

afterEach(async () => {
  testCluster.options.islands = [{ no: 0, nodes: [TEST_CONFIG] }];
  testCluster.options.shardsDiscoverIntervalMs = 1000000;
  testCluster.options.shardsDiscoverRecheckIslandsIntervalMs = 50;
  await testCluster.rediscover();
});

test("Shard is not discoverable error when connection got stuck", async () => {
  const STUCK_CONFIG = {
    ...TEST_CONFIG,
    host: connStuckServer.address().address,
    port: connStuckServer.address().port,
    connectionTimeoutMillis: 5000,
  };
  testCluster.options.islands = [
    {
      no: 0,
      nodes: [
        { ...STUCK_CONFIG, isAlwaysLaggingReplica: false, nameSuffix: "0" },
        { ...STUCK_CONFIG, isAlwaysLaggingReplica: true, nameSuffix: "0" },
      ],
    },
    {
      no: 1,
      nodes: [{ ...STUCK_CONFIG, nameSuffix: "1" }],
    },
  ];

  const shard = testCluster.shardByNo(0);

  const error1 = await join([
    jest.advanceTimersByTimeAsync(120000),
    shard.client(MASTER),
  ])
    .then(() => null)
    .catch((error: Error) => error.toString());
  expect(error1?.replace(/\d{3,}/g, "NNN")).toMatchSnapshot();

  const error2 = await shard
    .client(MASTER)
    .then(() => null)
    .catch((error: Error) => error.toString());
  expect(error2?.replace(/\d{3,}/g, "NNN")).toMatchSnapshot();
});

test("Shard is not discoverable error when no such shard", async () => {
  const shard = testCluster.shardByNo(1000);
  const error = await shard
    .client(MASTER)
    .then(() => null)
    .catch((error) => error.toString() as string);
  expect(error?.replace(/\d{3,}/g, "NNN")).toEqual(
    "ShardIsNotDiscoverableError: Shard NNN is not discoverable: no such Shard in the Cluster? some Islands are down? connections limit?",
  );
});

test("batcher() memoizes", async () => {
  const client = await testCluster.shardByNo(0).client(MASTER);

  const schemaA = new PgSchema("test", { id: { type: ID } }, []);
  const schemaB = new PgSchema("test", { id: { type: Number } }, []);
  const PgRunnerInsert = new PgQueryInsert(schemaA, { id: "42" }).RUNNER_CLASS;

  const batcher1 = client.batcher(
    PgQueryInsert,
    schemaA,
    "",
    false,
    () => new PgRunnerInsert(schemaA, client.client),
  );
  const batcher2 = client.batcher(
    PgQueryInsert,
    schemaA,
    "",
    false,
    () => new PgRunnerInsert(schemaA, client.client),
  );
  expect(batcher1).toBe(batcher2);

  const batcher3 = client.batcher(
    PgQueryInsert,
    schemaB,
    "",
    false,
    () => new PgRunnerInsert(schemaB, client.client),
  );
  expect(batcher1).not.toBe(batcher3);
});
