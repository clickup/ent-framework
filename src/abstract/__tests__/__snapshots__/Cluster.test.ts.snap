// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shard is not discoverable error when connection got stuck 1`] = `
"ShardIsNotDiscoverableError: Shard 0 is not discoverable: no such Shard in the Cluster? some Islands are down? connections limit?
- TestPgClient(test-pool(replica=false)-0).shardNos (took NNN ms): Error: Connection terminated due to connection timeout
- TestPgClient(test-pool(replica=false)-1).shardNos (took NNN ms): Error: Connection terminated due to connection timeout
- TestPgClient(test-pool(replica=true)-0).shardNos (took NNN ms): Error: Connection terminated due to connection timeout"
`;

exports[`Shard is not discoverable error when connection got stuck 2`] = `
"ShardIsNotDiscoverableError: Shard 0 is not discoverable: no such Shard in the Cluster? some Islands are down? connections limit?
- TestPgClient(test-pool(replica=false)-0).shardNos (took NNN ms): Error: Connection terminated due to connection timeout
- TestPgClient(test-pool(replica=false)-1).shardNos (took NNN ms): Error: Connection terminated due to connection timeout
- TestPgClient(test-pool(replica=true)-0).shardNos (took NNN ms): Error: Connection terminated due to connection timeout"
`;
